### ISSUE TITLE
Fetch errors not working correctly with the useMutation React hook

### ISSUE THREAD

**Issue by Densz** [jaydenseric/apollo-upload-client#222]
When I use createUploadLink in my React app 
```
const httpLink = createUploadLink({
  uri: 'http://localhost:4000/graphql"',
});

const client = new ApolloClient({
  link: from([authMiddleware, httpLink]),
  cache: new InMemoryCache(),
});
```
I cannot catch errors thrown by Mutations using `useMutation()`
```
const onSubmit = (data) => {
    mutationLogin({ variables: data })
      .then // it works when the mutation doesn't throw
      .catch // never goes in it even when the mutation throw
```
And on the browser console I get the error below
```
createUploadLink.js:148 Fetch failed loading: POST "http://localhost:4000/graphql".
```
And when use HttpLink from `@apollo/client` it works perfectly fine
```
const httpLink = new HttpLink({
  uri: 'http://localhost:4000/graphql"',
});
```

**Comment by jaydenseric** [comment]
Are you sure that mutation promises are meant to reject and be caught like that due to `fetch` errors? When the operation status resolves you have access to `error`, also in the `useMutation` options you can provide `onError`:

https://www.apollographql.com/docs/react/api/react/hooks/#usemutation

I can't see any code example in the Apollo docs using `.catch()` on a mutation, but in their network error docs, they say errors are best handled in Apollo Links:

> When using Apollo Link, the ability to handle network errors is way more powerful. The best way to do this is to use the `@apollo/client/link/error` to catch and handle server errors, network errors, and GraphQL errors.
> — https://www.apollographql.com/docs/react/data/error-handling/#network-errors

Maybe you have really found a HTTP link bug?  I don't have time to investigate right now, but if the `apollo-upload-client` behaviour really is incorrect, this might be a place to start:

https://github.com/jaydenseric/apollo-upload-client/blob/aaa8feafd34840096ced7686a5d3bc4aaa3dbc37/src/public/createUploadLink.js#L203-L207

Are the concepts in that comment sound?

**Comment by Densz** [comment]
It might be not recommanded to use it that way, but it can also be caught with a `catch` after calling the mutation
But just to be sure I tried the recommanded way, it doesn't work on my setup
```javascript
const [mutationLogin] = useMutation(MUTATION, {
    onError: (error) => console.log(error) // It never enters here
});
```
or
```javascript
const [mutationLogin, { error }] = useMutation...
console.log(error); // Error is always undefined
```
It might be a configuration problem on my side, I have to investigate further.

But when I replace in here
https://github.com/jaydenseric/apollo-upload-client/blob/aaa8feafd34840096ced7686a5d3bc4aaa3dbc37/src/public/createUploadLink.js#L208-L214
by the code below (just like in `createHttpLink` https://github.com/apollographql/apollo-client/blob/21b12f182c6b7a39a2174eabb4866f0ac6c82ed1/src/link/http/createHttpLink.ts#L135-L171)
```javascript
if (error.name === 'AbortError')
    return;
if (error.result && error.result.errors && error.result.data) {
    observer.next(error.result);
}
observer.error(error);
```
I managed to get the same behavior as createHttpLink using `error`, `onError` from `useMutation` or even `catch` on the mutation call itself.
I have to make further test on the project but for the moment it seems to have the correct behavior.

Let me know if there is something I'm missing :)

**Comment by jaydenseric** [comment]
I've sacrificed contract work today to look into this.

When testing fetch errors (I also wrote some more tests locally regarding invalid `fetch` `uri` to be sure) there are no problems using `ApolloClient.query` (or `mutation`):

https://github.com/jaydenseric/apollo-upload-client/blob/4c72c995756c3cc6db4b523f9c8a3158a42e9970/test/public/createUploadLink.test.js#L732-L776

The tests in this package mostly use `ApolloClient` methods like `query`, there are no tests using Apollo’s React hooks. It would be a fair bit of work to add them.

Apollo test their HTTP link in isolation at a low level (observers), using the [`execute` function from `apollo-link`](https://www.apollographql.com/docs/link/#standalone):

https://github.com/apollographql/apollo-client/blob/d470c964db46728d8a5dfc63990859c550fa1656/src/link/http/__tests__/HttpLink.ts#L360-L372

Writing tests like that are not intuitive to me; they have a lot of knowledge about how Apollo Client / Link internals are supposed to work.

Probably there is a bug here relating to the observer interface, but promise based stuff is fine.
